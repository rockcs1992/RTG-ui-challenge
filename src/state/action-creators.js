import {
    ADD_ITEM_QUANTITY,
    SUBTRACT_ITEM_QUANTITY,
    REMOVE_CART_ITEM,
    ADD_PRODUCT_TO_CART
} from './constants'

export const addItemQuantity = (sku) => ({
    type: ADD_ITEM_QUANTITY,
    sku
})

export const subtractItemQuantity = (sku) => ({
    type: SUBTRACT_ITEM_QUANTITY,
    sku
})

export const removeCartItem = (sku) => ({
    type: REMOVE_CART_ITEM,
    sku
})

export const addProductToCart = (product) => ({
    type: ADD_PRODUCT_TO_CART,
    product
})
