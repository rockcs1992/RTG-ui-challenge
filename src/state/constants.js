export const ADD_ITEM_QUANTITY = 'cart/ADD_ITEM_QUANTITY'
export const SUBTRACT_ITEM_QUANTITY = 'cart/SUBTRACT_ITEM_QUANTITY'
export const REMOVE_CART_ITEM = 'cart/REMOVE_CART_ITEM'

export const ADD_PRODUCT_TO_CART = 'product/ADD_PRODUCT_TO_CART'
