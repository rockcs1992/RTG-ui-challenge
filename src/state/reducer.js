import {
    ADD_ITEM_QUANTITY,
    SUBTRACT_ITEM_QUANTITY,
    REMOVE_CART_ITEM,
    ADD_PRODUCT_TO_CART
} from './constants'
import produce from 'immer'

export const initialState = {
    cart: []
  }

const reducer = produce((draft = initialState, action) => {
    switch(action.type) {
        case ADD_PRODUCT_TO_CART: {
            const addedItem = {
                ...action.product,
                quantity: 1
            }
            const index = draft.cart.findIndex((item) => item.sku === action.product.sku)
            if (index !== -1) {
                draft.cart[index].quantity++
            } else {
                draft.cart.push(addedItem)
            }
            break
        }
        case REMOVE_CART_ITEM: {
            draft.cart = draft.cart.filter((item) => item.sku !== action.sku)
            break
        }
        case ADD_ITEM_QUANTITY: {
            draft.cart = draft.cart.map((item) => {
                if (item.sku === action.sku) {
                    return { ...item, quantity: item.quantity + 1 }
                }
                return item
            })
            break
        }
        case SUBTRACT_ITEM_QUANTITY: {
            const index = draft.cart.findIndex((item) => item.sku === action.sku)
            if (index !== -1) {
                const quantity = draft.cart[index].quantity
                if (quantity <= 1) draft.cart.splice(index, 1)
                else draft.cart[index].quantity--
            }
            break
        }
        default: return draft
    }
})

export default reducer