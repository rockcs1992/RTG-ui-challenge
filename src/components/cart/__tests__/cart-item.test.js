import React from 'react';
import CartItem from '../cart-item'
import { ADD_ITEM_QUANTITY, SUBTRACT_ITEM_QUANTITY, REMOVE_CART_ITEM } from '../../../state/constants'
import { render, fireEvent} from '@testing-library/react'

const mockDispatch = jest.fn();

jest.mock('react-redux', () => ({
    useDispatch: ()=> mockDispatch,
    useSelector: () => [{ quantity: 1, price: 2}]
}));

describe('cart-checkout', () => {
    const item = {
        image: 'testurl',
        title: 'test',
        price: 1,
        quantity: 1,
        sku: 'test01'
    }
    it ('should render correctly', () => {
        const tree = render(
            <CartItem item={item} />,
        );
        expect(tree).toMatchSnapshot()
    })
    it ('should call ADD_ITEM_QUANTITY on button click', () => {
        const { getByText } = render(
            <CartItem item={item} />,
        );
        fireEvent.click(getByText('+'))
        expect(mockDispatch).toHaveBeenCalledWith({
            sku: 'test01',
            type: ADD_ITEM_QUANTITY
        })
    })
    it ('should call SUBTRACT_ITEM_QUANTITY on button click', () => {
        const { getByText } = render(
            <CartItem item={item} />,
        );
        fireEvent.click(getByText('-'))
        expect(mockDispatch).toHaveBeenCalledWith({
            sku: 'test01',
            type: SUBTRACT_ITEM_QUANTITY
        })
    })
    it ('should call REMOVE_CART_ITEM on button click', () => {
        const { getByText } = render(
            <CartItem item={item} />,
        );
        fireEvent.click(getByText('remove'))
        expect(mockDispatch).toHaveBeenCalledWith({
            sku: 'test01',
            type: REMOVE_CART_ITEM
        })
    })
})