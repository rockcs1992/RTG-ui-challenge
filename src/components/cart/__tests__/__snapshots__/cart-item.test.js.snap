// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cart-checkout should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="product cell small-12 grid-x grid-margin-x"
      >
        <img
          alt="test"
          src="testurl"
        />
        <div
          class="product-title cell small-4"
        >
          test
        </div>
        <div
          class="product-price cell small-2"
        >
          $
          1
        </div>
        <button
          class="product-price cell small-1 cart-quantity"
        >
          -
        </button>
        <div
          class="product-price cell small-1 cart-quantity"
        >
          1
        </div>
        <button
          class="product-price cell small-1 cart-quantity"
        >
          +
        </button>
        <button
          class="cart-remove"
        >
          remove
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="product cell small-12 grid-x grid-margin-x"
    >
      <img
        alt="test"
        src="testurl"
      />
      <div
        class="product-title cell small-4"
      >
        test
      </div>
      <div
        class="product-price cell small-2"
      >
        $
        1
      </div>
      <button
        class="product-price cell small-1 cart-quantity"
      >
        -
      </button>
      <div
        class="product-price cell small-1 cart-quantity"
      >
        1
      </div>
      <button
        class="product-price cell small-1 cart-quantity"
      >
        +
      </button>
      <button
        class="cart-remove"
      >
        remove
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
